protocol/messenger.go

```diff
@@ -2043,24 +2044,60 @@ func (m *Messenger) Mailservers() ([]string, error) {
 	return nil, ErrNotImplemented
 }
 
-func (m *Messenger) initChatFirstMessageTimestamp(chat *Chat) error {
-	if !chat.CommunityChat() || chat.FirstMessageTimestamp != FirstMessageTimestampUndefined {
-		return nil
+func (m *Messenger) initChatsFirstMessageTimestamp(chats []*Chat) {
+	var communityChats []*Chat
+	var communityChatIDs []string
+	for _, chat := range chats {
+		if !chat.CommunityChat() || chat.FirstMessageTimestamp != FirstMessageTimestampUndefined {
+			continue
+		}
+		communityChats = append(communityChats, chat)
+		communityChatIDs = append(communityChatIDs, chat.ID)
+	}
+	if len(communityChatIDs) == 0 {
+		return
 	}
 
-	oldestMessageTimestamp, hasAnyMessage, err := m.persistence.OldestMessageWhisperTimestampByChatID(chat.ID)
+	oldestMessageTimestamps, err := m.persistence.OldestMessageWhisperTimestampByChatIDs(communityChatIDs)
 	if err != nil {
-		return err
+		m.logger.Warn("failed to get oldest message timestamp", zap.Error(err))
+		return
 	}
+	communityCache := make(map[string]*communities.Community)
 
-	if hasAnyMessage {
-		if oldestMessageTimestamp == FirstMessageTimestampUndefined {
-			return nil
+	for _, chat := range communityChats {
+		community, ok := communityCache[chat.CommunityID]
+		if !ok {
+			community, err = m.communitiesManager.GetByIDString(chat.CommunityID)
+			if err != nil {
+				m.logger.Warn("failed to get community", zap.Error(err), zap.String("communityID", chat.CommunityID))
+				continue
+			}
+			communityCache[chat.CommunityID] = community
+		}
+
+		oldestMessageTimestamp, ok := oldestMessageTimestamps[chat.ID]
+		if !ok {
+			if err = m.updateChatFirstMessageTimestampForCommunity(chat, FirstMessageTimestampNoMessage, community); err != nil {
+				m.logger.Warn("failed to init first message timestamp", zap.Error(err), zap.String("chatID", chat.ID))
+			}
+			continue
+		}
+
+		if oldestMessageTimestamp != FirstMessageTimestampUndefined {
+			if err = m.updateChatFirstMessageTimestampForCommunity(chat, whisperToUnixTimestamp(oldestMessageTimestamp), community); err != nil {
+				m.logger.Warn("failed to init first message timestamp", zap.Error(err), zap.String("chatID", chat.ID))
+			}
 		}
-		return m.updateChatFirstMessageTimestamp(chat, whisperToUnixTimestamp(oldestMessageTimestamp), &MessengerResponse{})
 	}
 
-	return m.updateChatFirstMessageTimestamp(chat, FirstMessageTimestampNoMessage, &MessengerResponse{})
+	for _, community := range communityCache {
+		err := m.communitiesManager.SaveCommunity(community)
+		if err != nil {
+			m.logger.Warn("failed to save community", zap.Error(err), zap.String("communityID", community.IDString()))
+		}
+		m.communitiesManager.Publish(&communities.Subscription{Community: community})
```

Comment:
I don't think it makes sense to publish here, unless you actively checked if something changed. This way, we'll just be spamming the network on every all open.

Either you should check the community really changed and then publish, or just leave the publish to the normal sending loop.

If you go for the first option, you should also use `saveAndPublish` instead. It does the saving and the publishing while also checking that you are control node.